### PR TITLE
convert counter back to list #33229

### LINKED ISSUE #33229: MultiStepLR is broken with .step(epoch=xyz)
## 🐛 Bug

MultiStepLR is broken with `.step(epoch=xyz)`:

```
import torch
t = torch.tensor(1.0, requires_grad=True)
opt = torch.optim.SGD([t], lr=0.01)
s = torch.optim.lr_scheduler.MultiStepLR(opt, milestones=[19])
print('lr', opt.param_groups[0]['lr'])

opt.step()
s.step(0)
print('lr', opt.param_groups[0]['lr'])
```

Outputs:
```
lr 0.01
lr 0.001
```

I'm aware that `.step(epoch=xyz)` is deprecated, but it's used by third-party libraries, like https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L341.

cc @vincentqb @williamFalcon ^

I've debugged the issue and it seems to be caused by `bisect_right()` not playing well with `Counter`: https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L399

This can be fixed by casting self.milestones to list before bisection.  If that's an acceptable fix, I'm happy to send a PR.

## Environment

```
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 4.8.5-4ubuntu8) 4.8.5
CMake version: version 3.12.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.3

Versions of relevant libraries:
[pip3] numpy==1.17.2
[pip3] pytorch-lightning==0.5.3.2
[pip3] torch==1.4.0
[pip3] torch2trt==0.0.3
[pip3] torchvision==0.6.0a0
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by vincentqb**
Fixes #33229

### CODE REVIEW

**Comment by alsrgv** [comment]
Any updates to this PR?  Looking forward to seeing this merged.

**Comment by vincentqb** [comment]
As mentioned [here](https://github.com/pytorch/pytorch/issues/33229#issuecomment-589180630), a better long term solution may be to remove the dependency on `collections.Counter` to make serialization easier.

**Review by zhangguanheng66 — COMMENTED** [review]
Are users going to expect a BC breaking due to this fix?

**Comment by vincentqb** [comment]
> Are users going to expect a BC breaking due to this fix?

No.

In the case where the user provides an epoch parameter (which is being deprecated), there was an incorrect value being provided by MultiStepLR after #26423. This PR corrects this.

**Comment by ezyang** [comment]
Is the reason you're not doing the long term fix suggested in the issue? BC breakage?

**Review by ezyang — APPROVED** [review]
I'm hoping that people are not feeding a large list of milestones here, otherwise this is going to get expensive.

**Comment by vincentqb** [comment]
This modification is only needed for the closed form of the scheduler, which is being deprecated.

> I'm hoping that people are not feeding a large list of milestones here, otherwise this is going to get expensive.

Would something like #34325 be more efficient to you? Instead of converting `milestones` to a list at each iteration, `milestones` is processed as an iterator at each iteration.

If we want to avoid having to process at each iteration, we'd could convert to a list once. Let's call it `_milestones`. There's a risk that the counter `milestones` and the list `_milestones` would become out of sync. I prefer avoiding that risk, especially since we are deprecating the closed form.

Thoughts?

In the meanwhile, I'll continue with merging this PR.
